src/components/views/voip/VideoLobby.tsx

```diff
@@ -0,0 +1,208 @@
+    let contextMenu;
+    if (menuDisplayed) {
+        const selectDevice = (device: MediaDeviceInfo) => {
+            setDevice(device);
+            closeMenu();
+        };
+
+        const buttonRect = buttonRef.current.getBoundingClientRect();
+        contextMenu = <IconizedContextMenu {...aboveLeftOf(buttonRect)} onFinished={closeMenu}>
+            <IconizedContextMenuOptionList>
+                { labelledDevices.map(d =>
+                    <IconizedContextMenuOption
+                        key={d.deviceId}
+                        label={d.label}
+                        onClick={() => selectDevice(d)}
+                    />,
+                ) }
+            </IconizedContextMenuOptionList>
+        </IconizedContextMenu>;
+    }
+
+    if (!devices.length) return null;
+
+    return <div
+        className={classNames({
+            "mx_VideoLobby_deviceButtonWrapper": true,
+            "mx_VideoLobby_deviceButtonWrapper_active": active,
+        })}
+    >
+        <AccessibleTooltipButton
+            className={`mx_VideoLobby_deviceButton mx_VideoLobby_deviceButton_${kind}`}
+            title={active ? activeTitle : inactiveTitle}
+            alignment={Alignment.Top}
+            onClick={toggle}
+        />
+        { labelledDevices.length > 1 ? (
+            <ContextMenuButton
+                className="mx_VideoLobby_deviceListButton"
+                inputRef={buttonRef}
+                onClick={openMenu}
+                isExpanded={menuDisplayed}
+                label={deviceListLabel}
+            />
+        ) : null }
+        { contextMenu }
+    </div>;
+};
+
+const VideoLobby: FC<{ room: Room }> = ({ room }) => {
+    const [connecting, setConnecting] = useState(false);
+    const me = useMemo(() => room.getMember(room.myUserId), [room]);
+    const videoRef = useRef<HTMLVideoElement>();
+
+    const devices = useAsyncMemo(async () => {
+        try {
+            return await navigator.mediaDevices.enumerateDevices();
+        } catch (e) {
+            logger.warn(`Failed to get media device list: ${e}`);
+            return [];
+        }
+    }, [], []);
+    const audioDevices = useMemo(() => devices.filter(d => d.kind === "audioinput"), [devices]);
+    const videoDevices = useMemo(() => devices.filter(d => d.kind === "videoinput"), [devices]);
+
+    const [selectedAudioDevice, selectAudioDevice] = useState<MediaDeviceInfo>(null);
+    const [selectedVideoDevice, selectVideoDevice] = useState<MediaDeviceInfo>(null);
+
+    const audioDevice = selectedAudioDevice ?? audioDevices[0];
+    const videoDevice = selectedVideoDevice ?? videoDevices[0];
+
+    const [audioActive, toggleAudio] = useStateToggle(true);
+    const [videoActive, toggleVideo] = useStateToggle(true);
+
+    useEffect(() => {
+        if (videoDevice && videoActive) {
+            const videoElement = videoRef.current;
+            let stream: MediaStream;
+
+            (async () => {
```

Comment:
this could do with a comment to explain why we're sending it into an async closure

Reply:
hm, seems that this could stand to be split into a `useAsyncMemo` + synchronous `useEffect` anyways